Wait for hotplugged device

amshinde opened the pull request.

amshinde committed:
agent: Wait for device to be available before mounting it.

Hotplugged device may not be available within the VM at the
time of mouting it. Wait for the device to become available
by listening on udev events for block devices.

Fixes #75

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

amshinde committed:
vendor: Update vendoring to add package providing udev bindings

This package will be used for listening on udev events.
This will be used to check when devices that are hotplugged
become available within the VM.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

jcvenegas commented on syscall.go:
```diff
@@ -111,6 +114,62 @@ func unmountShareDir() error {
+	select {
+	case <-doneListening:
+		return
+	case <-time.After(1 * time.Second):
```
Could you create a constant for the device timeout ?

amshinde replied:
@jcvenegas I dont see the need to create a constant for this, I am using this value just once.

sboeuf replied:
You should remove `1 *` since you want 1 second.

I think we could wait up to 3 seconds, and in this case, the code should be:
```go
case <-time.After(time.Duration(3) * time.Second):
```
to avoid linter errors.

grahamwhaley replied:
@amshinde sometimes a constant is just nice for making it clearer to the reader what has been chosen/set and why - it doesn't have to hinge on the 'multi use' case.
Sure, it is a balance act, and in this case this is a pretty small file. I don't think it is a blocker.

amshinde replied:
@sboeuf Duration of 1 second is more than enough, if the device is not available by then means something has gone wrong.

amshinde committed:
ci: Install libudev-dev package

This is required for cgo dependency go-udev. Travis
environment does not have this installed.

Signed-off-by: Archana Shinde <archana.m.shinde@intel.com>

sboeuf commented on Gopkg.lock:
```diff
@@ -63,6 +63,18 @@
+
+[[projects]]
+  branch = "master"
+  name = "github.com/jochenvg/go-udev"
```
Please make sure we force the constraint to the master branch in the `Gopkg.toml` for this repo because we import it on purpose and we want to stick the latest version.
About `"github.com/jkeiser/iter"`, I guess this is a dependency pulled by the new package we have imported. That's okay not to constraint this one because we are not using it directly.

Always make sure you run
```
dep ensure
dep prune
```
after you have modified `Gopkg.toml`

sboeuf commented on .ci/setup.sh:
```diff
@@ -41,6 +41,9 @@ pushd "${runtime_repo_dir}"
 popd
 
+# Install libudev-dev required for go-udev vendor dependency
+sudo apt-get install libudev-dev
```
This commit can be removed since this is installed from the `tests` repo. https://github.com/clearcontainers/tests/pull/337 just got merged.

sboeuf replied:
Nevermind, this commit is completely fine, and we have to remove the one from the `tests` repo since it does not build the agent.

sboeuf commented on syscall.go:
```diff
@@ -111,6 +114,62 @@ func unmountShareDir() error {
+func waitForBlockDevice(deviceName string) (err error) {
+	devicePath := filepath.Join(devPath, deviceName)
+
+	if _, err := os.Stat(devicePath); err == nil {
```
Are we sure that because the device is here, it will be ready ?
I am asking because this would mean that we could call into `Stat()` from an infinite loop, until we either find the path, or hit the timeout.

grahamwhaley replied:
I guess that is true. If we do go that route, obviously put a small nap/sleep in the loop so we don't hard spin on it.

amshinde replied:
Yes, successful stat on the device would mean the device is ready. I went with the route of udev events as I find it a much cleaner approach, rather than polling continuously.

sboeuf replied:
I definitely agree it is cleaner, thanks for the confirmation.

sboeuf commented on syscall.go:
```diff
@@ -111,6 +114,62 @@ func unmountShareDir() error {
+		return err
+	}
+
+	// Create done signal channel for signalling epoll loop on the monitor socket.
```
s/signalling/signaling/

amshinde replied:
Tend to follow British English :)

sboeuf replied:
Oh! my bad...

sboeuf commented on syscall.go:
```diff
@@ -111,6 +114,62 @@ func unmountShareDir() error {
+		if _, err := os.Stat(devicePath); err == nil {
+			agentLog.Infof("Device %s already hotplugged, quit listening", deviceName)
+			close(done)
+			close(doneListening)
```
Please put a return at the end of this statement, this would allow avoiding the `else` statement which is not needed.

sboeuf commented on syscall.go:
```diff
@@ -111,6 +114,62 @@ func unmountShareDir() error {
+				agentLog.Infof("Event:", d.Syspath(), d.Action())
+				if d.Action() == "add" && filepath.Base(d.Devpath()) == deviceName {
+					agentLog.Infof("Hotplug event received for device %s", deviceName)
+					close(doneListening)
```
Why are you not closing `done` channel here ?

sboeuf reviewed (changes requested):
Few changes needed. And also need clarification about `Stat()` vs `netlink`. I guess that's because Stat() would be a polling loop, and netlink is more like an interrupt, right ?

amshinde commented:
@sboeuf Changes pushed.

sboeuf commented on syscall.go:
```diff
@@ -111,6 +114,61 @@ func unmountShareDir() error {
+	case <-doneListening:
+		close(done)
+	case <-time.After(time.Duration(1) * time.Second):
+		close(done)
```
Please error in this case.

amshinde replied:
Done

sboeuf commented on Gopkg.toml:
```diff
@@ -93,3 +93,7 @@
+
+[[constraint]]
+  branch = "master"
+  name = "github.com/jochenvg/go-udev"
```
Great, thanks !

sboeuf reviewed (commented):
Only one comment about returning an error in case of timeout. Other than that this PR sounds ready to be merged.
